### PR TITLE
new loading state tracker

### PR DESCRIPTION
**Pull request by thiagoc7**
Use isSaving and isFetching instead of ajaxCount.

If you guys like this solution, we need to think a good way to set a `minimum delay` before setting `loading state` back to false.

If it's too fast, like on localhost, it's not a good UX.

Thanks.

### CODE REVIEW

**Comment by justin808** [comment]
@alexfedoseev How does this look for you?

**Comment by alexfedoseev** [comment]
Looks good! :+1: 

According to minimum delay. It can smth like this:

```js
const minimumDelay = 500;

function handleAddCommentSuccess(dispatch, comment) {
  dispatch({
    type: COMMENT_ADD_SUCCEED,
    comment,
  });
}

export function addComment(comment) {
  return dispatch => {
    const startTime = (new Date()).getTime();
    dispatch({
      type: COMMENT_ADD_REQUESTED,
    });

    apiCall({ params })
      .then(res => {
        const comment = res.data;
        const successTime = (new Date()).getTime();
        const timeDiff = successTime - startTime;
        if (timeDiff < minimumDelay) {
          setTimeout(
            handleAddCommentSuccess.bind(null, dispatch, comment),
            minimumDelay - timeDiff
          );
        } else {
          handleAddCommentSuccess(dispatch, comment);
        }
      })
      .catch(res => {
        // same here...
      });
  }
}
```

We can wrap it in a nice helper.

**Comment by thiagoc7** [comment]
Thanks @alexfedoseev!

I just [changed](http://arguments.callee.info/2008/11/10/passing-arguments-to-settimeout-and-setinterval/) a a little the `setTimeout`, so we can avoid dispatch as an argument.

File: client/server.js [hunk 1]
```diff
@@ -35,7 +34,6 @@ server.app.get('/comments.json', function(req, res) {
 server.app.post('/comments.json', function(req, res) {
   console.log('Processing comment: %j', req.body.comment);
   console.log('(shhhh...napping 1 seconds)');
-  sleep.sleep(1);
```
Comment (justin808) [inline client/server.js]: @thiagoc7 Why did you remove this?

File: client/app/reducers/commentsReducer.js [hunk 2]
```diff
@@ -20,12 +21,14 @@ export default function commentsReducer($$state = $$initialState, action = null)
       return $$state.merge({
         $$comments: comments,
         fetchCommentError: null,
+        isFetching: false,
```
Comment (justin808) [inline client/app/reducers/commentsReducer.js]: We also show an ajax indicator when saving the record.

If that is the case, why call the flag `isFetching`? A reference to the fetch API?

How about `ajaxBusy`?

File: client/app/actions/commentsActionCreators.js [hunk 3]
```diff
@@ -43,24 +45,34 @@ export function submitCommentFailure(error) {
 
 export function fetchComments() {
   return dispatch => {
+    dispatch(setIsFetching());
     return (
-      commentsManager
-        .fetchComments()
-        .then(res => dispatch(fetchCommentsSuccess(res.data)))
-        .catch(res => dispatch(fetchCommentsFailure(res.data)))
+        commentsManager
```
Comment (justin808) [inline client/app/actions/commentsActionCreators.js]: why the extra indentation?

In general, try to limit whitespace diffs unless the linter is complaining.

Maybe the linter wants extra indentation?

File: client/app/actions/commentsActionCreators.js [hunk 4]
```diff
@@ -43,24 +45,34 @@ export function submitCommentFailure(error) {
...
+          const result = res.data;
+          const successTime = (new Date()).getTime();
+          const timeDiff = successTime - startTime;
+          if (timeDiff < minimumDelay) {
+            setTimeout(dispatch, minimumDelay - timeDiff, submitCommentSuccess(result));
+          } else {
+            dispatch(submitCommentSuccess(result));
+          }
+        })
         .catch(res => dispatch(submitCommentFailure(res.data)))
```
Comment (justin808) [inline client/app/actions/commentsActionCreators.js]: I'm a bit confused on setting a minimum delay on the client.

I had put in a sleep on the server to simulate a somewhat slow connection.

**Comment by justin808** [comment]
Hi @thiagoc7, I had a few questions for you. Thanks for contributing this.

File: client/server.js (hunk 1 above)
Comment (thiagoc7) [inline client/server.js]: @justin808, I removed because we don't need this if we the minimum delay func.
Let's talk about this below.

File: client/app/reducers/commentsReducer.js (hunk 2 above)
Comment (thiagoc7) [inline client/app/reducers/commentsReducer.js]: ;)

File: client/app/actions/commentsActionCreators.js (hunk 4 above)
Comment (thiagoc7) [inline client/app/actions/commentsActionCreators.js]: The idea of minimum delay is just for UX.
If the promise return very fast, you just have a blink, and it's kind of confusing.
But I agree that it's too much code for a starter repo...

**Comment by thiagoc7** [comment]
Hey @justin808, do you want me to remove this `minimum delay` stuff?
I also think that it's much more clear the way it were, with the `sleep` at server file.

I'll try to create a middleware for this, just don't know how yet... I'll take a look at tutorials this weekend.

File: client/app/actions/commentsActionCreators.js (hunk 4 above)
Comment (justin808) [inline client/app/actions/commentsActionCreators.js]: There's no need for middleware if we're just going put back the sleep in the server. The middleware was for the idea of having the action start a timeout for the processing indicator. I agree with you that it's just too much to get into having this extra code in the submit action. It's already complicated enough with the thunk middleware to do the asynch save.

File: client/app/actions/commentsActionCreators.js (hunk 3 above)
Comment (thiagoc7) [inline client/app/actions/commentsActionCreators.js]: don't know what happened, it's fixed

**Comment by justin808** [comment]
@alexfedoseev ready to merge?

**Comment by alexfedoseev** [comment]
Looks good! :+1:

**Comment by justin808** [comment]
@robwise Let's merge this.
